### DISCUSSION TITLE
Make foreign key constraint errors go through OnApplyChangesFailed interceptor

### DISCUSSION THREAD

**Discussion by slagtejn** [Mimetis/Dotmim.Sync#783] [Ideas]
I know I have asked something similar before, but probably did not explain it well.  In our syncing environment we have a scope called user data.  This scope contains all the tables needed for the users data with all the correct foreign key relations.  This scope is filtered, but how we handle filtered scopes is to create an instance of the base scope with a single filter parameter.  The user logging into the application can be linked to multiple users and thus will generate multiple user scopes.  For example the scopes would be in the format

UserData_1
UserData_2
UserData_3

We do this because it's a dynamically changing parameter and we would need to get the historical data of a user if they are linked to a new user.  A manager getting a new employee for example.  The scope_info table will contain the timestamp for each filter.  With this implementation we don't have to re-initialize the scope every time the parameter changes to get the historical data and it's transferring a lot less data if new users are added instead of getting all data for all linked users.  Over time some of the scopes will become obsolete and the data in the database becomes old, so we have a mechanism to clean it up.

Ok so now the problem.  There are cases were foreign key data is crossing scopes.  For example some data from UserData_1 crosses into UserData_2.  When this happens there are foreign key constraint errors and sync just stops with no other data syncing on the scope.   Now UserData_2 syncs and the data is inserted so when UserData_1 syncs again it might finish.  If there are a lot of FK errors this process would take a long time and could end in an endless loop.  An endless loop would be a data issue that needs to be fixed on server, but at least sync would be able to detect it instead of failing.

Sync framework was able to handle this with the knowledge blob which contained the failed records and this error went through OnApplyChangedFailed interceptor with the type 'ErrorsOccurred'.  Sync was able to continue syncing the rest of the data for the scope, but ignore those failed records somehow (still a mystery on how it was able to do that).

Do you see a possible solution to this?  I know disabling foreign keys will work, but that locks the table and is not the best solution since it could ruin the integrity of the data. I can think of adding a table that contains the json of the failed records per scope and appends them during the next sync of a specific scope so data is not lost.  If sync continues and updates the timestamp on the scope then the failed records would not be retrieved again so there needs to be some sort of tracker.

**Comment by Mimetis** [comment]
My first guess would be to use `DisableConstraintsOnApplyChanges` (in `SyncOptions`) to allow any data to be synced without failing because of foreign keys constraints.

For the errors occured, it's tricky.
I don't see exactly what kind of errors can be raised in such situation.
Can you share a sample where this situation can occurs ?

**Comment by slagtejn** [comment]
Yes disabling the foreign keys will work, but the tables are locked during sync and data inconsistencies can happen.  In most scenarios sync is a background process so now any application using the database will have queries hanging because sync has locks on the tables within the scope.

I made a sample a few months ago.  https://github.com/slagtejn/Dotmim.Sync.FKError

This issue is not limited to this specific scenario.  It's a general issue when using filters on a relational database.  The data has to be perfect for sync to finish.  If filtered data is crossing tables then sync could continue to fail instead of continuing to sync the rest of the data and only fail specific records.  It will fail at the same point every sync cycle and now it's in a broken state with no way to fix it.

Yes it's tricky to find a solution for this.

**Reply by Mimetis** [reply]
Yes, I saw your sample.
Took me a long time to understand it, without any context, it was quite complicated.

I've rewrite it using a more "human readable" way :)
Here is my version:

``` cs
var serverDbName = "FServer";
var clientDbName1 = "Employee1";
var clientDbName2 = "Employee2";

await DBHelper.CreateDatabaseAsync(serverDbName, true);
await DBHelper.CreateDatabaseAsync(clientDbName1, true);
await DBHelper.CreateDatabaseAsync(clientDbName2, true);

var script = @"
CREATE TABLE Customer (CustomerId int IDENTITY(5000, 1000) NOT NULL PRIMARY KEY, Name varchar(50) Not Null, EmployeeId int NOT NULL);

CREATE TABLE Sales (SalesId int IDENTITY(100, 10) NOT NULL PRIMARY KEY, EmployeeId int NOT NULL, BuyerCustomerId int NOT NULL, Product varchar(50) NOT NULL,
CONSTRAINT FK_Buyer_Customer FOREIGN KEY(BuyerCustomerId) REFERENCES Customer(CustomerId));

SET IDENTITY_INSERT Customer ON
INSERT Customer (CustomerId, [Name], EmployeeId) VALUES(5000, 'B. Gates', 1)
INSERT Customer (CustomerId, [Name], EmployeeId) VALUES(6000, 'S. Nadela', 1)
INSERT Customer (CustomerId, [Name], EmployeeId) VALUES(7000, 'S. Balmer', 1)
INSERT Customer (CustomerId, [Name], EmployeeId) VALUES(8000, 'S. Jobs', 2)
INSERT Customer (CustomerId, [Name], EmployeeId) VALUES(9000, 'T. Cook', 2)
SET IDENTITY_INSERT Customer OFF

INSERT Sales (EmployeeId, BuyerCustomerId, Product) VALUES (1, 5000, 'Stairs');
INSERT Sales (EmployeeId, BuyerCustomerId, Product) VALUES (1, 6000, 'Doors');
INSERT Sales (EmployeeId, BuyerCustomerId, Product) VALUES (2, 8000, 'Oranges');
-- We have a problem here. An employee 1 sold something to a customer that is not in its customers list 
-- Customer 9000 is affiliated to employee 2
INSERT Sales (EmployeeId, BuyerCustomerId, Product) VALUES (1, 9000, 'Strawberries');
";

await DBHelper.ExecuteScriptAsync(serverDbName, script);

var progress = new SynchronousProgress<ProgressArgs>(s =>
{
    Console.ForegroundColor = ConsoleColor.Green;
    Console.WriteLine($"{s.ProgressPercentage:p}:  \t[{s.Source[..Math.Min(4, s.Source.Length)]}] {s.TypeName}:\t{s.Message}");
    Console.ResetColor();

});

var serverProvider = new SqlSyncProvider(DBHelper.GetDatabaseConnectionString(serverDbName));
var employee1Provider = new SqlSyncProvider(DBHelper.GetDatabaseConnectionString(clientDbName1));
var employee2Provider = new SqlSyncProvider(DBHelper.GetDatabaseConnectionString(clientDbName2));

var setup = new SyncSetup("Customer", "Sales");
setup.Filters.Add("Customer", "EmployeeId");
setup.Filters.Add("Sales", "EmployeeId");

try
{

    var emp1Agent = new SyncAgent(employee1Provider, serverProvider);
    var emp1Params = new SyncParameters(("EmployeeId", 1));
    var emp1result = await emp1Agent.SynchronizeAsync(setup, emp1Params, progress);
    Console.WriteLine(emp1result);

}
catch (Exception ex)
{
    Console.WriteLine(ex);
}

```

Basically, in this sample, the `Employee 1` will sell something to a customer that is not affiliated with him.
And of course, since the customer is not existing in its DB, it will fail.

The problem we have here is that we are using a `Merge` statement, and `Merge` is not able to make a "_continue on error_" statement.
It's all or nothing.

For 4 lines merged, if 3 are ok, and 1 is failing, the whole 4 lines would come in errors....

**Comment by slagtejn** [comment]
Yes I think we only use the Bulk stored procs and the setting 'UseBulkOperations' is never used.  Maybe if these records fail we fallback to use the _update stored proc and try to insert them one at a time so then we know which records failed.

**Comment by VagueGit** [comment]
We have this issue a lot. We cannot disable foreign keys as sync errors can result in inconsistent data. So we often find a single sync error can cause all subsequent syncs to fail.

Our customers are beyond our control so we have customers who ignore sync errors for months at a time. `OutdatedAction.ReinitializeWithUpload` fails due to the sync error so the customer can lose months of data.

**Reply by Mimetis** [reply]
If you have any suggestions to by pass foreign keys  without disabling them, let me know :D

**Reply by slagtejn** [reply]
I think what we are trying to get at is to have a way to track failed records so the entire sync process does not stop.  Obviously this is a lot harder said than done and would have to come up with a nice solution.  Right now if there are any errors in the process sync will just stop and there are cases like FK errors where it can continue if handled correctly.

**Comment by VagueGit** [comment]
Perhaps I have is understood. We don't want to bypass foreign keys. We would prefer if a record that failed to sync did not cause all subsequent data to fail to sync.

I think where it gets tricky is in determining which row has the error that causes a batch to fail. Is there any guidance on how to do that? If we had that information we could remediate.

**Reply by Mimetis** [reply]
That's the problem... the `MERGE` statement is not able to continue on error and log any row in error
It would be possible I guess, if all `insert` / `update` would have been done 1 by 1, but the performances would become a disaster.

Each batch is by default **10 000** rows (`CoreProvider.BulkBatchMaxLinesCount`) 
So the best we could be is to say "If 1 error occurs in a batch (from 1 to 10 K rows) we are raising an error for the entire batch"
And try to continue.

Here is a sample scenario using the `AdventureWorks` database:

Imagine we have 1 error in `ProductCategory`, it will then raise an error for all `ProductCategory` (since it's only 1 batch, with around 20 lines)
Then, if we allow to "`Continue On Error`", **DMS** will try to insert rows for `Product`.

But since no `ProductCategory` rows have been inserted, all the `Product` rows will fail too... 
And so on for `Sales` , `SalesOrderHeader` .....

At the end almost all tables failed to sync.
But DMS will commit the whole sync, since we allowed to "`Continue On error`" 
And now, the problems begin :D

**Reply by slagtejn** [reply]
I am not sure performance would be that bad and it's better than having the process fail.  Before sync framework 2.1 bulk procedures didn't exist.  There was a performance increase when upgrading, but it wasn't huge.

I know I keep mentioning sync framework, but it had scenarios that it was able to magically handle like this FK issue.  The sync process never stopped and it was able to insert/update/delete all the records besides the failed ones.  On every subsequent sync it would keep trying to apply the failed records and in some cases it was an endless loop, but we were able to see which records were constantly failing and fix the data. 

There needs to be a way to track the failed records.  If the sync process is allowed to "Continue On Error" then these records would never be retrieved again (unless there was an update to them) because the scope would be updated to the latest timestamp.  Without a failed record tracker it's not possible to implement.

**Comment by VagueGit** [comment]
How about an option to choose 1 record at a time? We service lots of companies but they typically sync 1 order at a time. Bulk operations aren't that important for us.

At the moment even if we choose a very small batch size, we still see hundreds of orders in a batch.

**Comment by Mimetis** [comment]
I've just create a POC that re-introduce the possibility to choose between bulk or not bulk operations for **SQL Server** (**Sqlite** & **MySql** are not concerned here, since bulk is not supported) using the `UseBulkOperations` property, available on `SqlSyncProvider`.

It's a major performance downgrade, especially for first sync (when initializing a new client database)
But I guess you can play with this value depending on the situation...

The  I've created a new interceptor called (I think I will rename it ... if you have any idea ;) ) `OnApplyChangesErrorOccured`.
Implementation is quite easy: IF an error occurs, you have the row in error, and you need to pass an `ErrorResolution` (`Continue` or `Throw`) value

``` cs
var serverProvider = new SqlSyncProvider(DBHelper.GetDatabaseConnectionString(serverDbName));

var employee1Provider = new SqlSyncProvider(DBHelper.GetDatabaseConnectionString(clientDbName1))
{
    // .... And .... We're back again
    UseBulkOperations = false
};

try
{
    var emp1Agent = new SyncAgent(employee1Provider, serverProvider);
    var emp1Params = new SyncParameters(("EmployeeId", 1));

    emp1Agent.LocalOrchestrator.OnApplyChangesErrorOccured(args =>
    {
        Console.WriteLine(args.ErrorRow);
        // We can do something here the failed row
        // ....
        // Then pass the resolution to Continue to prevent a fail 
        args.Resolution = ErrorResolution.Continue;
    });

    var emp1result = await emp1Agent.SynchronizeAsync(setup, emp1Params, progress);
    Console.WriteLine(emp1result);
}

```
![image](https://user-images.githubusercontent.com/4592555/184258261-9d707b7f-d372-44fd-95d8-b00e99626863.png)

As you can see we downloaded 6 rows, but only 5 have been applied locally. 
My only concern here is that now, the client database is not synced with the server, since we missed a row...

Your thoughts ?

**Reply by Mimetis** [reply]
You have the branch https://github.com/Mimetis/Dotmim.Sync/tree/OnErrorOccured to test it.

100 % sure it will not stay as is.
I think I will try to merge the `OnApplyChangesErrorOccured` with the `OnApplyChangesFailed` 
Not sure, I need to check if it's faisable

But I would like to have only 1 single interceptor for everything related to **Conflicts** or **Errors** during an apply changes.

**Reply by slagtejn** [reply]
Do you ever sleep?  That was fast.  Yes now the data is not in-sync with the server which is fine for this session, but in the next session it will never try to apply that failed record again.  We need a table to track the failed records so on the next sync it would try to apply them again.  What do you think?  I am not sure what would be stored there.  We could store the entire JSON of the row or maybe just scope_id, scope_name, PK of the record, tablename?  Also would need to handle the case of the failed record being updated and always take the latest version of changes.

I agree it should be within the OnApplyChangesFailed interceptor with the type ErrorsOccured (I think it's already there but not used).

**Reply by Mimetis** [reply]
2 AM right now. I go sleep :D

**Reply by Mimetis** [reply]
> We need a table to track the failed records so on the next sync it would try to apply them again

Not possible. The apply changes on the client is the last operation. 
If we try to do that, it means `DMS` needs to send back the record to the server and say "hey it's not done yet, mark this row has not updated"

The problem is that someone could have update the row at the same time, from another client, or even the server itself

Another problem would be the infinite error. If we mark the error as "not updated", the row will be downloaded again and again over time leading to an infinite error state. (Maybe not the case for Foreign Key error, but for sure for some others errors)
The example we have in this thread will actually leads to an infinite error loop, until someone changes the data in error.

Also storing a row in error will involve creating a new `sync_scope_errors` table on the server, and I don't want more tables that what we have today.

By the way, if you want to call back the server to mark the row as not updated, you can create your own mechanism / web api to do so.
You have the full row in error in the `OnApplyChangesErrorOccured`.

**Reply by slagtejn** [reply]
Sounds good.  We will implement our own solution to handle the failed records.  Will the interceptor have the exception incase we want to do something specific based on the type of exception?

Edit:  I should have looked at the POC first.  I see the exception as part of the args

**Comment by slagtejn** [comment]
Will this solution work when UsingBulkOperations? If an error occurs on the bulk records it falls back to using the single stored proc to figure out the failed records.  We would be able to UseBulkOperations and still figure out which records failed by the fallback

**Comment by VagueGit** [comment]
I think it is ok for bulk operations to be the default. Then on error switch to row-at-a-time-sync and then sync again to find the faulting row.

After a faulting row has been remediated, switch back to bulk operations. This might be done in a loop ... switch back and forth between bulk operations until the last faulting row is remediated.

**Reply by Mimetis** [reply]
I guess it can be a solution :D

In my opinion, the bulk mode is useful for the first sync, where you are downloading a lot of rows
From that point, I guess line per line should be enough.

Of course, in a regular scenario like a mobile application ...

Remember that bulk mode is only implemented in the Sql Server Sync Provider. it's not existing for MySQL / MariaDB / SQLite

**Reply by slagtejn** [reply]
We have scenarios when clients are offline for long periods of time.  When coming back online they sync millions of records so bulk operations do help.  I like the solution of falling back to row-at-a-time-sync on error so bulk operations can still be used (at least for SQL server).

**Comment by slagtejn** [comment]
Will this be finished in 0.9.6? I noticed it was in the draft, but does it work when using bulk operations?

**Reply by Mimetis** [reply]
I'm still searching a solution. For now only "row per row" is working fine.
For bulk operations, today, all the batch is raising the error.

Let me explain the situation:

The entire batch is failing. Batch size is provided by `SyncProvider` (default to 10k rows `public virtual int BulkBatchMaxLinesCount { get; set; } = 10000;`)
A batch is containing 1 and only 1 table: If a table (that should be applied on server or client) is contained in a single batchfile, another batch is created for the next table to be applied.
So far, if one row is failing in a batch containing a table with less than 10K rows, the entire table is failing.... 
And you will have a crazy amount of `OnApplyChangesErrorOccured` interceptor raised, since this interceptor is raised for each error...

That's being said, I still have an idea, but I don't know yet if it's feasible:

_IF a batch is failing THEN fallback to line by line, for this batch only_

I need to see if it's feasible. 
What do you think ?

**Reply by Mimetis** [reply]
@VagueGit @slagtejn 

I've implemented the solution in the current dev branch [ClientHistory](https://github.com/Mimetis/Dotmim.Sync/tree/ClientHistory)

If you can test it and make feedback would be cool :)

**Reply by slagtejn** [reply]
I tested this and it's working great!  I know you were not a fan of a solution like this, but it really makes DMS more robust and extensible to anyone that uses this library.

**Reply by Mimetis** [reply]
Hey @slagtejn 
I see that you have a deep experience on the old Sync Framework.

Since you have used both framework, do you have any feature in mind that is not supported by DMS (and still present in Sync Framework) ?

**Reply by slagtejn** [reply]
Sync Framework had the ability to track failed records and sync them again over and over.  The OnApplyChangedFailed interceptor is a great improvement, but when you implement the the error resolution of 'Continue' that failed record will never download again.  Sync framework tracked this within the knowledge blob at the scope level so it was able to download or upload the failed records the next time a scope synced.  I am not sure how it did this, but I assume it used the '_selectrow' stored proc.  Currently DMS has this stored proc, but I don't think it's used.  It also joins the tracking table which is not necessary since not all records in DMS will have tracking data.  In sync framework every record had tracking data and it would insert into the _tracking tables on provision (would take forever!). 

With DMS we are now able to have multiple filter parameter values per scope, but you will find out that data has to line up perfectly for it to sync in a relational database.  There are many cases where the parameter values cross and will error out because of foreign key errors.  This might be ok because on the next sync of the same scope with a different parameter value could insert the missing data.  When syncing the original scope with the foreign key error again the record would be inserted.

Of course there are cases when an error can never be resolved and happen over/over again because of data issues, but at least that could be seen with logging and DMS does it's job of trying to sync the record instead of skipping it.  If the record is skipped and never synced it's very hard to figure out what happened months later.

It would be nice to have some sort of failed record tracking and include them when selecting changes.  I know you don't want more tables on the server, but I am not sure how else to do this unless there is a blob on the scope that contains the failed records.  DMS would have to merge them together when selecting changes.  The database record takes priority incase there are changes and the failed record is out of date.

If you want a new table maybe something like below and you can use the _selectrow stored proc to get the data to join with _changes.

scope_info_errors
	sync_scope_id
	sync_scope_name
	record_id
	table_name
	error_message
        isupload ??? not sure about this.  table could exist in both client/server so schema's match

**Reply by slagtejn** [reply]
Sync framework was also able to pick up where it left off when using smaller transactions. It updated the scope_info and knowledge blob with the minimum server/client timestamps. I don’t know exactly how sync framework figured it out.

DMS it’s now possible to commit changes per batch which is a very nice feature, but if the entire process fails in-between it will have to download everything again.

It would be nice to pickup where it left off and the only way to do that would be to track the row version per table.  Obviously large changes to DMS so this is a nice to have.

**Reply by slagtejn** [reply]
Sync framework had the ability to sync a scope with a direction. If the scope is bi-directional you could specify to only upload changes. We used this feature when loggin off the application as a final sync to upload changes.

**Reply by Mimetis** [reply]
> DMS will have tracking data. In sync framework every record had tracking data and it would insert into the _tracking tables on provision (would take forever!)

Earlier version of `DMS` worked like that, but I decided to change this behavior later on. 
It simplified the first sync, as we do not have to fill an entire tracking table anymore, as you said.

> but I assume it used the '_selectrow' stored proc. Currently DMS has this stored proc, but I don't think it's used. 

It's used when you call the `GetConflictAsync()` in the interceptor `OnChangesConflictOccured` (new version of `OnApplyChangesFailed`)

> It would be nice to have some sort of failed record tracking and include them when selecting changes

Indeed, could be a good idea.
Something we can imagine, that will not require to change everything, would be to:

- If a row is failing or is in conflict.
- AND you specify a new error / conflict resolution, like:
   - For conflict, let's say : `ConflictResolution.RetryOnNextSync`
   - For error, let's say : `ErrorResolution.RetryOnNextSync`
- The primary key row is kept in memory until the end of the sync
- All the error's row are sent back to the server (or client) and timestamp is updated to be part of the next sync again

Benefits: 
- It does not require to change anything in the scopes tables
- It allows a new scenario where you are deciding to download rows gain later on

Cons:
- If you have a lot of failing rows, it will increase the memory consumption on the client (storing a large amount of primary keys datas)
   - Of course we can mitigate this by storing the primary keys in a batch
- It requires one more step at the end of the sync process.
- It will not be transactional (even if the mode `TransactionMode.AllOrNothing` is selected)

I think it's a really nice behavior to have and I will work on it for the next version (`v0.9.7`)
In my opinion, it will not require any complicated upgrade on both sides since it's only more code to add (and no schema changes)

> It would be nice to pickup where it left off and the only way to do that would be to track the row version per table. 

Easy to do with `SQL Server` where a row timestamp is available. Way too difficult for `MySql `& `Sqlite`.
I can't see a simple way to implement it in a future version of `DMS`, to be honest.

> Sync framework had the ability to sync a scope with a direction

You can specify a direction for each table, directly from your `SyncSetup` instance:

``` cs
 var setup = new SyncSetup("Employee", "Address", "EmployeeAddress", "Product", "ProductCategory");
 setup.Tables["Employee"].SyncDirection = SyncDirection.Bidirectional;  // Default value
 setup.Tables["Address"].SyncDirection = SyncDirection.DownloadOnly;
 setup.Tables["ProductCategory"].SyncDirection = SyncDirection.UploadOnly;
 setup.Tables["Product"].SyncDirection = SyncDirection.UploadOnly;
```

@slagtejn , once again, thanks a ton for all the feedbacks you made during this summer time, to improve DMS. 
Your knowledge and expertise on sync is really valuable, and it helped me a lot.
DMS is way more robust and new greats features have been implemented so far.

Now, it's time to stabilize the engine. 
I think I will (maybe) implement the `RetryOnNextSync` feature, but it will be the last major upgrade to DMS until the version `v1.0`.

Thanks a ton

**Reply by slagtejn** [reply]
Nice!  Didn't know the '_selectrow' stored proc was used, but makes sense now.

I like you will try to implement the failed record tracking since that is the most impactful feature.  Another con of the in-memory approach is if sync stops in-between (application closes) , then the failed records are lost.  Are you sure you don't want to persist this data?   You could store it as a file instead of inside the database.  Would the timestamp be updated on the actual record or is this an in-memory timestamp?  I don't think you can update timestamps in SQL Server.  Would need to change something on the record.  If the timestamp is updated on the actual record then all other syncing clients would download the change.

I should have clarified the ability to sync a scope with a direction.  You could specify the direction during runtime.  If a scope was configured to sync bi-directional you could specify you only want to upload data.  Sync framework didn't have the ability to sync tables within  a scope differently, but in DMS you can control per table.

**Reply by Mimetis** [reply]
@slagtejn 

Just validate the new beta version of `v0.9.6` with the errors handling.
Need your feedback on this new feature, explained here : https://github.com/Mimetis/Dotmim.Sync/discussions/802#discussioncomment-3593987

**Reply by slagtejn** [reply]
Does this only work with AllOrNothing transaction mode?  I get the below error because it's only creating a connection if the transaction mode is AllOrNothing.  I think the code needs to be

if (Options.TransactionMode == TransactionMode.AllOrNothing || Options.TransactionMode == TransactionMode.PerBatch)

![image](https://user-images.githubusercontent.com/97198525/189172599-3543564a-7f8a-4fa0-b940-87e0167fe16a.png)

**Reply by Mimetis** [reply]
No it's a bug, I need to make all tests with different transaction mode, I guess.
Will fix asap

**Reply by slagtejn** [reply]
I validated the error handling using RetryOnNextSync and it works great!  I like the new sync_scope_errors column containing meta data to the physical files on disk.  It keeps the database clean and fast since no need to query a large table containing the errors.  In my opinion it is the best feature in this release.  It really makes DMS robust!!

Few comments

1. If a record fails for a few sync cycles, the 'ERRORS_' folder is not being cleaned up.  It will create a new folder on every sync with the same data.  It should be feasible to delete the old folder since it's contained in the sync_scope_errors column.  If the folder is changing then delete the old one.  Also when the record succeeds the files & folder (if no more errors) are not cleaned up.  Would this be possible?
2. You changed the 'SetParameterValue' method in BaseOrchestrator.Command.cs to be 'InternalSetParameterValue'.  We have built tables on top of DMS that contain stats/device/user information and we used the 'SetParamterValue' function to fill the tables.  Can this be changed back or create a new public one called SetParameterValue that just calls the internal one?  There is now not a way to use this function since it's internal.  Or move it out to some static DB helper class because these functions are static

**Reply by slagtejn** [reply]
There is also an error when trying to apply the errors during upload.  Moving creating BatchInfo to the top of RemoteOrchestrator.ApplyChanges solved the issue, but the changes are still not uploaded. 

```csharp

                // Errors batch info
                BatchInfo errorsBatchInfo = null;

                try
                {
                    // Create a batch info for error rows on "retry apply errros" or "apply changes":
                    string info = connection != null && !string.IsNullOrEmpty(connection.Database) ? $"{connection.Database}_ERRORS" : "ERRORS";
                    errorsBatchInfo = new BatchInfo(this.Options.BatchDirectory, info: info);

```

![image](https://user-images.githubusercontent.com/97198525/189200923-a788d77b-a00c-49e5-a519-63e5a59e9d7a.png)

**Reply by Mimetis** [reply]
I've made some modification on the **HandlingErrors** branch.
If you can take a look on it, would be cool

Thanks @slagtejn, all the tests and feedbacks you've done so far, helps me so much (and makes `DMS` better day after day)
**Very Much Appreciated**

**Reply by slagtejn** [reply]
I tried the latest branch and it's working better now!  I can see the folders are being cleaned up on both the server/client depending on upload/download errors.  Also the records are being updated correctly if they are fixed.  There is a scenario that is not working yet, but I am not sure how to fix it unless the entire scope succeeds.  If the entire scope succeeds then the sync_scope_error column and associated files should be deleted.  Here is the scenario...

1. Force a record to fail by deleting the foreign key constraint on the server and putting in a value that doesn't exist in the client.
2. Sync will flag it as an error (FK constraint) and the sync_scope_error column will be updated and an ERROR file will be created.
3. Put back the foreign key constraint and update the record on the server to a value that exists in the client.
4. Sync again and it will try to apply the error record, but it will still fail because it has old data.  The data is not merged with the newly changed record from the server.  Maybe this will solve the issue?
5. Sync will apply the new record and this time it will succeed, but the corresponding error record will not be cleaned up.
6. Sync again and it will try to apply the old failed record now forever.

It is not possible to fix the sync error by updating data on either the server/client unless because the records are not merged.  It could also apply outdated data to a record because the errors records are not updated.

**Reply by Mimetis** [reply]
Before Sync again, try to delete the folder, and let me know what happens.
If it's not working, I will need a demo with AdventureWorks to be able to reproduce, fix, and create the test

**Reply by slagtejn** [reply]
That will work for this scenario, but then it will never apply a failed record if the folder is always deleted.  If the failed record is never updated on the server, but data from another table is updated that causes the failed record to be inserted then it should be inserted.  This record would come from the ERROR folder.  Even if it was weeks later.

The scenario that doesn't work is if the same failed record is updated, it will apply the error record first and then apply the updated record.  The error record would fail and the updated record would succeed, but the error record would never be cleaned up and keep trying to apply over and over again.  If the entire scope succeeds the error records should be cleaned up and it would solve some of the issue.  It would not solve if 1 record is still failing in the scope.

**Reply by Mimetis** [reply]
Your scenario depends on a schema changed (remove and add again a constraint)
It's an edge scenario, and I'm not sure `DMS` can do something about it

**Reply by slagtejn** [reply]
Yeah the scenario is just an easy way to replicate the issue and it is a scenario that won't happen too often.  The idea is if something on the failing record is updated that causes it to succeed (changing a value that is used in a filter), then the old error record will not be cleaned up try to apply again.  It could succeed this time, but it would apply the old data or it could fail over and over even though a new version of the record was already inserted.  

I'm not sure how to solve this since there is not a link between the failed record and a possible update to the same record unless they are merged based on the PK so always the latest record is used.

We have a process to cleanup the batch dir so this issue will eventually fix itself

**Reply by Mimetis** [reply]
I've implemented something in the last commit.

On Apply Changes phase, IF an error batch exists, I get all error rows and I check if the row has been updated locally with a timestamp that is newer than the last sync.
If yes, that means the line has been updated since the error has been recorded and saved locally
Then I delete the row.

That means the issue is resolved by itself and will not happens again.
This especially useful if you are using a `RetryOnNextSync` value

Let me know if it solves your scenario

**Reply by slagtejn** [reply]
Wow you always come up with a solution! I believe this will solve the issue. I will let you know shortly. Yes I was using the RetryOnNextSync value

**Reply by slagtejn** [reply]
I tried the solution and it worked the first time, but if you make that same record fail again it will clean up the error on the next sync which is not correct. It should continue to fail until the record is changed to succeed.

**Reply by Mimetis** [reply]
This errors handling is giving me headaches..
It's way more difficult that what I was expecting...

I've made a bunch of changes in the code branch [HandlingErrors](https://github.com/Mimetis/Dotmim.Sync/tree/HandlingErrors) trying to handle all situations.

Can you give it a try and let me know what you think ?

**Reply by slagtejn** [reply]
Well it's working better, but I still encounter a scenario that doesn't work because the scope_last_sync_timestamp on the scope is not updated until the next sync. 

1. Make a record fail
2. Sync
3. Fix the record
4. Sync
5. The scope_last_sync_timestamp is not updated.
6. Make the same record fail again
7. Sync will clean it up because the timestamp on the record > scope_last_sync_timestamp

If you do step 4 twice it will work because the scope_last_sync_timestamp would be updated.

![image](https://user-images.githubusercontent.com/97198525/189712265-2ffef753-3be2-4916-9b69-2f0c366ad02d.png)

**Reply by Mimetis** [reply]
Where the record is failing ? on client or server ?
When you Sync, which Retry do u use ? `RetryOnNextSync` I guess ?
Your screenshot is coming from which side ? Server / CLient ?

Are you able to reproduce the error easily  ? If yes, are you able to write a small markdown error code ?

**Reply by slagtejn** [reply]
Everything is happening on the client. Screenshot is from the client. It’s easy to replicate. Just update a record on the server and the scope_last_sync_timestamp will not be updated until the second sync.

I use the RetryOnNextSync option.

**Reply by Mimetis** [reply]
What kind of error are you able to generate on the server that will fail on the client ?

**Reply by Mimetis** [reply]
Ok I got a scenario

**Reply by Mimetis** [reply]
Rhaaaaa it's tough !

Ok, let's give it a new try, 
I tried a new technic, relying on changes coming, more than on tracking tables metadata

(As Always on the HandlingErrors branch)

I think it's almost good !

**Reply by slagtejn** [reply]
Yes it's working now!  I tested download/upload failed records along with conflict handling.  I like that it's now merging the changes together based on the PK.  Seems a lot safer than relying on the timestamps and it will always take the latest version of the row.

**Reply by Mimetis** [reply]
Eventually it's working !! :)
Good, I guess the `v0.9.7` **Release** is going to be published soon :)

**Reply by Mimetis** [reply]
I've added a default option called `ErrorResolutionPolicy` that you can set in your `SyncOptions` instance (like the `ConflictResolutionPolicy` that we already have)
It will allow you to set the default error policy that will be used for all errors, until you are using the `OnApplyChangesConflictOccured` interceptor to change it for a particular table / row

And by the way, the doc is also done : https://dotmimsync.readthedocs.io/Errors.html (*Yeeeeeeeeee*)

**Reply by slagtejn** [reply]
Very nice work!  Looks good and works good.  Best feature of the release.

**Comment by VagueGit** [comment]
Can a similar scenario arise without a schema change?
Parent record deleted on one client and synced to server
On another client, parent delete has failed to sync and so attempts to sync child records ....

I see that quite a bit in our logs.

Maybe I'm barking up the wrong branch again.

 **EDIT:** It's such a big issue that I might have to introduce soft deletes in the next release of this product ... with all the querying overhead that requires

**Reply by Mimetis** [reply]
Try with `ContinueOnError`, and the sync will end without failing
Error row will then be stored in the tmp folder
